vocabulary-draft/continuous-delivery.md

```diff
@@ -0,0 +1,114 @@
+
+- **TaskRun Started**: a TaskRun inside a PipelineRun has started.
+- **TaskRun Finished**: a TaskRun inside a PipelineRun has finished. 
+
+Pipeline Events MUST include the following attributes:
+- **Event Type**: the type is restricted to include `CD.Pipeline**` prefix. For example `CD.Pipeline.PipelineRun.Queued` or `CD.Pipeline.Tests.Started`
+- **PipelineRun Id**: unique identifier for a pipeline execution
+- **Pipeline Name**: unique identifier for the pipeline, not for the instance. A pipeline can have multiple instances/runs.  
+- **PipelineRun Status**: current status of the PipelineRun at the time when the event was emitted. If the pipeline is finished, this attribute should reflect if it finished successfully or if there was an error on the execution. Possible statuses: [Running, Finished, Error]
+
+Optional attributes: 
+- **PipelineRun URL**: URL to locate where  pipeline instances are running
+- **PipelineRun Errors**: error field to indicate possible compilation, test, build and package errors.
+
+## Product Composition Events
+
+When the products delivered to users can be, or needs to be, broken down into components (hardware, software, configuration files etc.), the creation of versions of said components and the compositoning of components into products or product subsystems may warrant the use of separate events detailed in this section.
+
+Product Composition Events MUST include the following attributes:
+
+- **Event Type**: the type is restricted to include `CD.ProductComposition` prefix. For example `CD.ProductComposition.ComponentVersionCreated` or `CD.ProductComposition.CompositionCreated`
+
+### Component Version Created
+
+Announces that a new version of a Product Component has been created.
+
+Examples of Component Versions include:
+
+- New version of a software module or application.
+- New version of a document or piece of configuration.
+- New revision of a hardware component.
+
+The purpose of this event would be for systems to start creating Product Compositions / Baseline Versions that include this Component Version.
+
+Component Version Created Events MUST include the following attributes:
+
+- **Component Version ID**: Unique identifier of component version.
+- **Component ID**: Identifier of component of which this is a version.
+
+Optional attributes:
+
+- **Version Name**: Machine-friendly and human-friendly version tag/number/etc. for component version, e.g. “1.4.2-alpha2”.
+- **Component Name**: Machine-friendly and human-friendly name of component, e.g. “c37-backbone-router”.
+- **Component Display Name**: UX/human-friendly display name of component, e.g. “C37 Backbone Router”.
+- **Artifacts**: List of artifacts (files, packages, Docker images etc.) part of this component version.
+- **Component Track**: States what track (e.g. platform variant (x86_64, ARM etc.), project (e.g. X-DataCenter, X-Cloud etc.) or other) the component, of which this is a version, belongs to.
+
```

Comment:
@erkist I am confused.. at the begining of this section it is stated as example: 
"- New version of a document or piece of configuration."

But then a component has a list of **Artifacts**: List of artifacts (files, packages, Docker images etc.) part of this component version.

Reply:
@salaboy Hmm, maybe that can be clarified somehow. In my view, a new version of a document can either be represented by a file (which would then be an artifact) or it could just be an entity in some type of service (in which case there is no file to talk about, and thus presumably no artifact?). The same goes for a piece of configuration, it can also be either a file or not a file. And then we have hardware, which is never a file.

Now that I think about it, maybe I have a too strict connection between Artifact and File in my head? How do you see Artifacts?

Reply:
@erkist as soon as that is version controlled and you can change it it is an artifact. For me configuration files are artifacts as soon as they are stored in version control, the same with documents. Then you have binary artifacts that have a version associated to them. Hence I tend to think that Hardware can be classified as artifact for example: raspberry pi version 3.1 with serial number XXX, how do you link to those artifacts might be with a reference number that is outside of the system. If you are planning to deploy something on top of a specific hardware I think that you can consider the hardware itself as an environment, let's say: `deploy binary X to raspberry pi environment` my two cents..

Reply:
@salaboy If things that are not files can be artifacts, and we can add "arbitrarily complex metadata" to them, there is no need for Component Versions.

If. however, we want to see artifacts only as "a binary object under version control", then they will not solve my needs and I will need component versions.